include/nbl/builtin/hlsl/cpp_compat/impl/intrinsics_impl.hlsl

```diff
@@ -3,428 +3,573 @@
+	}
+};
+template<typename Vector>
+requires concepts::FloatingPointVector<Vector>
+struct length_helper<Vector>
 {
-	static vector<int32_t, N> findMSB(NBL_CONST_REF_ARG(vector<int32_t, N>) val)
+	static inline typename vector_traits<Vector>::scalar_type __call(NBL_CONST_REF_ARG(Vector) vec)
+	{
+		return std::sqrt(dot_helper<Vector>::__call(vec, vec));
+	}
+};
+template<typename Vectorial>
+requires concepts::FloatingPointLikeVectorial<Vectorial>
+struct normalize_helper<Vectorial>
+{
+	static inline Vectorial __call(NBL_CONST_REF_ARG(Vectorial) vec)
+	{
+		return vec / length_helper<Vectorial>::__call(vec);
+	}
+};
+template<typename T>
+requires concepts::Scalar<T>
+struct max_helper<T>
+{
+	static T __call(NBL_CONST_REF_ARG(T) a, NBL_CONST_REF_ARG(T) b)
+	{
+		return std::max(a, b);
+	}
+};
+template<typename T>
+requires concepts::Scalar<T>
+struct min_helper<T>
+{
+	static T __call(NBL_CONST_REF_ARG(T) a, NBL_CONST_REF_ARG(T) b)
+	{
+		return std::min(a, b);
+	}
+};
+template<typename T>
+requires concepts::IntegralScalar<T>
+struct find_lsb_helper<T>
+{
+	using return_t = int32_t;
+	static inline T __call(const T arg)
+	{
+		return glm::findLSB(arg);
+	}
+};
+template<typename Integer>
+NBL_PARTIAL_REQ_TOP(concepts::IntegralScalar<Integer>)
+struct find_msb_helper<Integer NBL_PARTIAL_REQ_BOT(concepts::IntegralScalar<Integer>) >
+{
+	using return_t = int32_t;
+	static return_t __call(NBL_CONST_REF_ARG(Integer) val)
 	{
-#ifdef __HLSL_VERSION
-		return spirv::findSMsb(val);
-#else
 		return glm::findMSB(val);
-#endif
 	}
 };
-
-#ifndef __HLSL_VERSION
-
 template<typename EnumType>
-	requires std::is_enum_v<EnumType>
+requires std::is_enum_v<EnumType>
+struct find_lsb_helper<EnumType>
+{
+	using return_t = int32_t;
+	static int32_t __call(NBL_CONST_REF_ARG(EnumType) val)
+	{
+		using underlying_t = std::underlying_type_t<EnumType>;
+		return find_lsb_helper<underlying_t>::__call(static_cast<underlying_t>(val));
+	}
+};
+template<typename EnumType>
+requires std::is_enum_v<EnumType>
 struct find_msb_helper<EnumType>
 {
-	static int32_t findMSB(NBL_CONST_REF_ARG(EnumType) val)
+	using return_t = int32_t;
+	static return_t __call(NBL_CONST_REF_ARG(EnumType) val)
 	{
 		using underlying_t = std::underlying_type_t<EnumType>;
-		return find_msb_helper<underlying_t>::findMSB(static_cast<underlying_t>(val));
+		return find_msb_helper<underlying_t>::__call(static_cast<underlying_t>(val));
 	}
 };
 
-#endif
-
-template<typename Integer>
-struct find_lsb_helper;
-
-template<>
-struct find_lsb_helper<int32_t>
+template<typename FloatingPoint>
+requires concepts::FloatingPointScalar<FloatingPoint>
+struct rsqrt_helper<FloatingPoint>
 {
-	static int32_t findLSB(NBL_CONST_REF_ARG(int32_t) val)
+	static FloatingPoint __call(NBL_CONST_REF_ARG(FloatingPoint) x)
 	{
-#ifdef __HLSL_VERSION
-		return spirv::findILsb(val);
-#else
-		return glm::findLSB(val);
-#endif
+		// TODO: https://stackoverflow.com/a/62239778
+		return 1.0f / std::sqrt(x);
 	}
 };
```

Comment:
maybe it would be slightly better do express `rsqrt_helper` in terms of `sqrt_helper` then you could relax the concept to just `FloatingPointLike` (emulated and native scalars and vectors)

Reply:
then we still would need [this](https://github.com/Devsh-Graphics-Programming/Nabla/blob/new_tgmath/include/nbl/builtin/hlsl/cpp_compat/impl/intrinsics_impl.hlsl#L533) but disabled for C++, so there is no ambigous specialization and it works for HLSL... will leave it as it is

Reply:
ok I see what you did later, all good